File: backend/controllers/authController.js
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";
import crypto from "crypto";
import User from "../models/User.js";
import {
  sendVerificationEmail,
  sendPasswordResetEmail,
} from "../utils/email.js";

export const generateToken = (id) => {
  return jwt.sign({ id }, process.env.JWT_SECRET, {
    expiresIn: "30d",
  });
};

import PlatformSettings from "../models/PlatformSettings.js";

export const signup = async (req, res) => {
  const { name, email, password } = req.body;

  try {
    // Check if registration is enabled
    const settings = await PlatformSettings.findOne();
    if (settings && !settings.allowNewRegistrations) {
      return res.status(403).json({ message: "Registration is currently disabled by Admin." });
    }

    const userExists = await User.findOne({ email });
    if (userExists) {
      return res.status(400).json({ message: "User already exists" });
    }

    const hashedPassword = await bcrypt.hash(password, 10);
    const verificationToken = crypto.randomBytes(32).toString("hex");

    const user = await User.create({
      name,
      email,
      password: hashedPassword,
      verificationToken,
      isVerified: process.env.NODE_ENV !== "production",
    });

    try {
      await sendVerificationEmail(email, verificationToken);
    } catch (err) {
      console.warn(
        `Email verification failed to send to ${email}, please use this link: http://localhost:3000/verify-email?token=${verificationToken}`,
      );
      console.error("Email error details:", err);
    }

    res.status(201).json({
      message:
        process.env.NODE_ENV !== "production"
          ? "User registered successfully. Status: Auto-verified (Development Mode)."
          : "User registered successfully. Please check your email to verify your account.",
    });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const login = async (req, res) => {
  const { email, password } = req.body;

  try {
    const user = await User.findOne({ email });
    if (!user) {
      return res.status(400).json({ message: "Invalid credentials" });
    }

    if (user.isBlocked) {
      return res.status(403).json({ message: "User blocked by Admin." });
    }

    if (!user.isVerified) {
      return res
        .status(401)
        .json({ message: "Please verify your email first" });
    }

    const isMatch = await bcrypt.compare(password, user.password);
    if (!isMatch) {
      return res.status(400).json({ message: "Invalid credentials" });
    }

    const token = generateToken(user._id.toString());

    res.cookie("token", token, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      maxAge: 30 * 24 * 60 * 60 * 1000, // 30 days
    });

    res.json({
      _id: user._id,
      name: user.name,
      email: user.email,
      bio: user.bio,
      avatar: user.avatar,
      avatar3d: user.avatar3d,
      avatarMode: user.avatarMode,
      skills: user.skills,
      title: user.title,
      settings: user.settings,
      links: user.links,
      skillswapId: user.skillswapId,
      country: user.country,
      education: user.education,
      experience: user.experience,
      achievements: user.achievements,
      platforms: user.platforms,
      followers: user.followers,
      following: user.following,
      rating: user.rating,
      numReviews: user.numReviews,
      completedSwaps: user.completedSwaps,
      token,
    });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const verifyEmail = async (req, res) => {
  const { token } = req.query;

  try {
    const user = await User.findOne({ verificationToken: token });
    if (!user) {
      return res
        .status(400)
        .json({ message: "Invalid or expired verification token" });
    }

    user.isVerified = true;
    user.verificationToken = undefined;
    await user.save();

    res.json({ message: "Email verified successfully. You can now log in." });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const requestPasswordReset = async (req, res) => {
  const { email } = req.body;

  try {
    const user = await User.findOne({ email });
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    const resetToken = crypto.randomBytes(32).toString("hex");
    user.resetPasswordToken = resetToken;
    user.resetPasswordExpires = new Date(Date.now() + 3600000); // 1 hour
    await user.save();

    await sendPasswordResetEmail(email, resetToken);

    res.json({ message: "Password reset link sent to your email." });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const resetPassword = async (req, res) => {
  const { token, newPassword } = req.body;

  try {
    const user = await User.findOne({
      resetPasswordToken: token,
      resetPasswordExpires: { $gt: new Date() },
    });

    if (!user) {
      return res
        .status(400)
        .json({ message: "Invalid or expired reset token" });
    }

    user.password = await bcrypt.hash(newPassword, 10);
    user.resetPasswordToken = undefined;
    user.resetPasswordExpires = undefined;
    await user.save();

    res.json({ message: "Password reset successfully. You can now log in." });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const getMe = async (req, res) => {
  try {
    const user = req.user;
    if (!user) {
      return res.status(401).json({ message: "Not authorized" });
    }

    // Update lastActive when user accesses their data (effectively "opens website/app")
    await User.findByIdAndUpdate(user._id, { lastActive: new Date() });

    res.json({
      _id: user._id,
      name: user.name,
      email: user.email,
      bio: user.bio,
      avatar: user.avatar,
      avatar3d: user.avatar3d,
      avatarMode: user.avatarMode,
      skills: user.skills,
      title: user.title,
      settings: user.settings,
      links: user.links,
      skillswapId: user.skillswapId,
      country: user.country,
      education: user.education,
      experience: user.experience,
      achievements: user.achievements,
      platforms: user.platforms,
      followers: user.followers,
      following: user.following,
      rating: user.rating,
      numReviews: user.numReviews,
      completedSwaps: user.completedSwaps,
      totalHours: user.totalHours,
      topSkills: user.topSkills,
      totalEndorsements: user.totalEndorsements,
      isVerified: user.isVerified,
    });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const logout = async (_req, res) => {
  res.cookie("token", "", {
    httpOnly: true,
    expires: new Date(0),
  });
  res.json({ message: "Logged out successfully" });
};
export const updateProfile = async (req, res) => {
  try {
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    const { name, bio, avatar, skills, title, avatar3d, avatarMode } = req.body;

    // Build update object explicitly to ensure all fields are captured
    const updateFields = {};
    if (name) updateFields.name = name;
    if (bio !== undefined) updateFields.bio = bio;
    if (avatar !== undefined) updateFields.avatar = avatar; // 2D Image Base64
    if (skills !== undefined) {
      // Sanitize: filter out any strings that look like MongoDB ObjectIds (24 hex characters)
      // to ensure only human-readable skill names are stored in the profile.
      updateFields.skills = Array.isArray(skills)
        ? skills.filter(skill => typeof skill === 'string' && !/^[0-9a-fA-F]{24}$/.test(skill))
        : skills;
    }
    if (title !== undefined) updateFields.title = title;
    if (avatar3d !== undefined) updateFields.avatar3d = avatar3d; // 3D Object
    if (avatarMode !== undefined) updateFields.avatarMode = avatarMode;
    if (req.body.settings !== undefined) updateFields.settings = req.body.settings;
    if (req.body.links !== undefined) updateFields.links = req.body.links;
    if (req.body.skillswapId !== undefined) updateFields.skillswapId = req.body.skillswapId;
    if (req.body.country !== undefined) updateFields.country = req.body.country;
    if (req.body.education !== undefined) updateFields.education = req.body.education;
    if (req.body.experience !== undefined) updateFields.experience = req.body.experience;
    if (req.body.achievements !== undefined) updateFields.achievements = req.body.achievements;
    if (req.body.platforms !== undefined) updateFields.platforms = req.body.platforms;

    const updatedUser = await User.findByIdAndUpdate(
      req.user._id,
      { $set: updateFields },
      { returnDocument: 'after', runValidators: true }
    );

    if (!updatedUser) {
      return res.status(404).json({ message: "User not found" });
    }

    res.json({
      _id: updatedUser._id,
      name: updatedUser.name,
      email: updatedUser.email,
      bio: updatedUser.bio,
      avatar: updatedUser.avatar,
      avatar3d: updatedUser.avatar3d,
      avatarMode: updatedUser.avatarMode,
      skills: updatedUser.skills,
      title: updatedUser.title,
      settings: updatedUser.settings,
      links: updatedUser.links,
      skillswapId: updatedUser.skillswapId,
      country: updatedUser.country,
      education: updatedUser.education,
      experience: updatedUser.experience,
      achievements: updatedUser.achievements,
      platforms: updatedUser.platforms,
      followers: updatedUser.followers,
      following: updatedUser.following,
      rating: updatedUser.rating,
      numReviews: updatedUser.numReviews,
      completedSwaps: updatedUser.completedSwaps,
      totalHours: updatedUser.totalHours,
      topSkills: updatedUser.topSkills,
      totalEndorsements: updatedUser.totalEndorsements,
    });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const changePassword = async (req, res) => {
  const { currentPassword, newPassword } = req.body;

  try {
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }

    const isMatch = await bcrypt.compare(currentPassword, user.password);
    if (!isMatch) {
      return res.status(400).json({ message: "Incorrect current password" });
    }

    const hashedPassword = await bcrypt.hash(newPassword, 10);
    user.password = hashedPassword;
    await user.save();

    res.json({ message: "Password updated successfully" });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

export const deleteAccount = async (req, res) => {
  try {
    await User.findByIdAndDelete(req.user._id);
    res.cookie("token", "", {
      httpOnly: true,
      expires: new Date(0),
    });
    res.json({ message: "Account deleted successfully" });
  } catch (error) {
    res.status(500).json({ message: "Server error", error });
  }
};

File: backend-spring/src/main/java/com/skillswap/backend/controller/AuthController.java
package com.skillswap.backend.controller;

import com.skillswap.backend.model.User;
import com.skillswap.backend.repository.PlatformSettingsRepository;
import com.skillswap.backend.repository.UserRepository;
import com.skillswap.backend.security.JwtTokenProvider;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PlatformSettingsRepository platformSettingsRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @PostMapping("/signup")
    public ResponseEntity<?> signup(@RequestBody Map<String, String> body) {
        String name = body.get("name");
        String email = body.get("email");
        String password = body.get("password");

        if (name == null || email == null || password == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Name, email and password are required"));
        }

        if (userRepository.findByEmail(email).isPresent()) {
            return ResponseEntity.badRequest().body(Map.of("message", "User already exists"));
        }

        String hashedPassword = passwordEncoder.encode(password);
        String verificationToken = generateToken();

        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPassword(hashedPassword);
        user.setVerificationToken(verificationToken);
        user.setVerified(true); // Auto-verify in dev
        user.setCreatedAt(LocalDateTime.now());
        user.setUpdatedAt(LocalDateTime.now());
        userRepository.save(user);

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(Map.of("message", "User registered successfully. Auto-verified (Development Mode)."));
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> body, HttpServletResponse response) {
        String email = body.get("email");
        String password = body.get("password");

        Optional<User> userOpt = userRepository.findByEmail(email);
        if (userOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid credentials"));
        }

        User user = userOpt.get();

        if (user.isBlocked()) {
            return ResponseEntity.status(403).body(Map.of("message", "User blocked by Admin."));
        }

        if (!user.isVerified()) {
            return ResponseEntity.status(401).body(Map.of("message", "Please verify your email first"));
        }

        if (!passwordEncoder.matches(password, user.getPassword())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid credentials"));
        }

        String token = jwtTokenProvider.generateToken(user.getId());

        Cookie cookie = new Cookie("token", token);
        cookie.setHttpOnly(true);
        cookie.setPath("/");
        cookie.setMaxAge(30 * 24 * 60 * 60);
        response.addCookie(cookie);

        Map<String, Object> result = buildUserResponse(user);
        result.put("token", token);
        return ResponseEntity.ok(result);
    }

    @GetMapping("/me")
    public ResponseEntity<?> getMe(@AuthenticationPrincipal User user) {
        if (user == null) return ResponseEntity.status(401).body(Map.of("message", "Not authorized"));
        // Update lastActive
        user.setLastActive(LocalDateTime.now());
        userRepository.save(user);
        return ResponseEntity.ok(buildUserResponse(user));
    }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(HttpServletResponse response) {
        Cookie cookie = new Cookie("token", "");
        cookie.setHttpOnly(true);
        cookie.setPath("/");
        cookie.setMaxAge(0);
        response.addCookie(cookie);
        return ResponseEntity.ok(Map.of("message", "Logged out successfully"));
    }

    @GetMapping("/verify-email")
    public ResponseEntity<?> verifyEmail(@RequestParam String token) {
        Optional<User> userOpt = userRepository.findAll().stream()
                .filter(u -> token.equals(u.getVerificationToken()))
                .findFirst();
        if (userOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid or expired verification token"));
        }
        User user = userOpt.get();
        user.setVerified(true);
        user.setVerificationToken(null);
        userRepository.save(user);
        return ResponseEntity.ok(Map.of("message", "Email verified successfully. You can now log in."));
    }

    @PostMapping("/request-password-reset")
    public ResponseEntity<?> requestPasswordReset(@RequestBody Map<String, String> body) {
        String email = body.get("email");
        Optional<User> userOpt = userRepository.findByEmail(email);
        if (userOpt.isEmpty()) {
            return ResponseEntity.status(404).body(Map.of("message", "User not found"));
        }
        User user = userOpt.get();
        String resetToken = generateToken();
        user.setResetPasswordToken(resetToken);
        user.setResetPasswordExpires(LocalDateTime.now().plusHours(1));
        userRepository.save(user);
        return ResponseEntity.ok(Map.of("message", "Password reset link sent to your email."));
    }

    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@RequestBody Map<String, String> body) {
        String token = body.get("token");
        String newPassword = body.get("newPassword");
        Optional<User> userOpt = userRepository.findAll().stream()
                .filter(u -> token.equals(u.getResetPasswordToken())
                        && u.getResetPasswordExpires() != null
                        && u.getResetPasswordExpires().isAfter(LocalDateTime.now()))
                .findFirst();
        if (userOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(Map.of("message", "Invalid or expired reset token"));
        }
        User user = userOpt.get();
        user.setPassword(passwordEncoder.encode(newPassword));
        user.setResetPasswordToken(null);
        user.setResetPasswordExpires(null);
        userRepository.save(user);
        return ResponseEntity.ok(Map.of("message", "Password reset successfully. You can now log in."));
    }

    @PutMapping("/profile")
    public ResponseEntity<?> updateProfile(@AuthenticationPrincipal User currentUser,
                                           @RequestBody Map<String, Object> body) {
        if (currentUser == null) return ResponseEntity.status(401).body(Map.of("message", "Not authorized"));
        User user = userRepository.findById(currentUser.getId()).orElse(null);
        if (user == null) return ResponseEntity.status(404).body(Map.of("message", "User not found"));

        if (body.containsKey("name")) user.setName((String) body.get("name"));
        if (body.containsKey("bio")) user.setBio((String) body.get("bio"));
        if (body.containsKey("avatar")) user.setAvatar((String) body.get("avatar"));
        if (body.containsKey("title")) user.setTitle((String) body.get("title"));
        if (body.containsKey("country")) user.setCountry((String) body.get("country"));
        if (body.containsKey("skillswapId")) user.setSkillswapId((String) body.get("skillswapId"));
        user.setUpdatedAt(LocalDateTime.now());
        userRepository.save(user);
        return ResponseEntity.ok(buildUserResponse(user));
    }

    @PutMapping("/change-password")
    public ResponseEntity<?> changePassword(@AuthenticationPrincipal User currentUser,
                                            @RequestBody Map<String, String> body) {
        if (currentUser == null) return ResponseEntity.status(401).body(Map.of("message", "Not authorized"));
        User user = userRepository.findById(currentUser.getId()).orElse(null);
        if (user == null) return ResponseEntity.status(404).body(Map.of("message", "User not found"));

        if (!passwordEncoder.matches(body.get("currentPassword"), user.getPassword())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Incorrect current password"));
        }
        user.setPassword(passwordEncoder.encode(body.get("newPassword")));
        userRepository.save(user);
        return ResponseEntity.ok(Map.of("message", "Password updated successfully"));
    }

    @DeleteMapping("/delete-account")
    public ResponseEntity<?> deleteAccount(@AuthenticationPrincipal User currentUser,
                                           HttpServletResponse response) {
        if (currentUser == null) return ResponseEntity.status(401).body(Map.of("message", "Not authorized"));
        userRepository.deleteById(currentUser.getId());
        Cookie cookie = new Cookie("token", "");
        cookie.setHttpOnly(true);
        cookie.setPath("/");
        cookie.setMaxAge(0);
        response.addCookie(cookie);
        return ResponseEntity.ok(Map.of("message", "Account deleted successfully"));
    }

    // Helper
    private Map<String, Object> buildUserResponse(User user) {
        Map<String, Object> result = new HashMap<>();
        result.put("_id", user.getId());
        result.put("name", user.getName());
        result.put("email", user.getEmail());
        result.put("bio", user.getBio());
        result.put("avatar", user.getAvatar());
        result.put("avatar3d", user.getAvatar3d());
        result.put("avatarMode", user.getAvatarMode());
        result.put("skills", user.getSkills());
        result.put("title", user.getTitle());
        result.put("settings", user.getSettings());
        result.put("links", user.getLinks());
        result.put("skillswapId", user.getSkillswapId());
        result.put("country", user.getCountry());
        result.put("education", user.getEducation());
        result.put("experience", user.getExperience());
        result.put("achievements", user.getAchievements());
        result.put("platforms", user.getPlatforms());
        result.put("followers", user.getFollowers());
        result.put("following", user.getFollowing());
        result.put("rating", user.getRating());
        result.put("numReviews", user.getNumReviews());
        result.put("completedSwaps", user.getCompletedSwaps());
        result.put("totalHours", user.getTotalHours());
        result.put("topSkills", user.getTopSkills());
        result.put("totalEndorsements", user.getTotalEndorsements());
        result.put("isVerified", user.isVerified());
        return result;
    }

    private String generateToken() {
        byte[] bytes = new byte[32];
        new SecureRandom().nextBytes(bytes);
        return HexFormat.of().formatHex(bytes);
    }
}
